### PR TITLE
Frogger-game

### PR DESCRIPTION
**Pull request by vladk96**
### Frogger-game

[Demo](https://vladk96.github.io/frogger-game/) |
[Code base](https://github.com/vladk96/frontend-2019-homeworks/blob/frogger-game/submissions/vladk96/frogger-game/js/app.js)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

**Comment by OleksiyRudenko** [comment]
@vladk96 Looks nice. Still there is space for improvements.

Please, check yourself thoroughly against the following list (e.g. requirements re Constants are implemented partially).

 * [ ] it is OK to employ ES6 features like `const`, `let` etc.
 * [ ] OO is implemented using JS prototype chain object model (**not** ES6 classes syntax)
 * Requirements re **Constants**:
   * [ ] all numbers like block dimensions, initial locations are defined as constants
   * [ ] there are core constants and derived constants
         (e.g. `const FIELD_WIDTH = BLOCK_WIDTH * BLOCKS_NUMBER;`)
   * [ ] arrays of constants are also constants 
         (e.g. `const INITIAL_POSITIONS = [1,2,3,4].map(rowNumber => rowNumber * BLOCK_HEIGHT);`)
   * [ ] const objects help organizing and structure const data even better
         (e.g. `const PLAYER_CONF = { initialPosition: {x: 1, y: 5}, sprite: '...', ...etc... };`
 * Requirements re **OOP**:
   * [ ] properties common for some classes are generalized into a base class
         (e.g. there is `Character` base class, which is extended by `Enemy` and `Player` classes)
   * [ ] class extension is implemented using `Subclass.prototype = Object.create(Superclass.prototype)`,
         not `Subclass.prototype = new Superclass(params);`;
         [Useful resource](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Inheritance_and_the_prototype_chain)
   * [ ] classes do not refer to any global variables, like global variable `player`, which is an instance of `Player` class
         (referring to global constants and globals provided by the gaming platform like `Resources` is OK);
         Hint: pass `Player` instance as an argument to every enemy
   * [ ] Separation of Concerns principle is followed
         (e.g. `update` method does only rendering and 
         doesn't **contain** any inline code to check e.g. collisions)
   * [ ] the code is very [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
 * Most common mistakes
   * [ ] Make sure `target = condition ? valueWhenConditionTrue : valueWhenConditionFalse` is used instead of
         `condition ? target = valueWhenConditionTrue : target = valueWhenConditionFalse`;
         [Conditional (ternary) operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator)

Please, confirm all items above are positive. Please, ask a question if anything above requires clarification.

**Comment by vladk96** [comment]
> @vladk96 Looks nice. Still there is space for improvements.
> 
> Please, check yourself thoroughly against the following list (e.g. requirements re Constants are implemented partially).
> 
> * [x]  it is OK to employ ES6 features like `const`, `let` etc.
> * [x]  OO is implemented using JS prototype chain object model (**not** ES6 classes syntax)
> * Requirements re **Constants**:
>   
>   * [x]  all numbers like block dimensions, initial locations are defined as constants
>   * [ ]  there are core constants and derived constants
>     (e.g. `const FIELD_WIDTH = BLOCK_WIDTH * BLOCKS_NUMBER;`)
>   * [ ]  arrays of constants are also constants
>     (e.g. `const INITIAL_POSITIONS = [1,2,3,4].map(rowNumber => rowNumber * BLOCK_HEIGHT);`)
>   * [ ]  const objects help organizing and structure const data even better
>     (e.g. `const PLAYER_CONF = { initialPosition: {x: 1, y: 5}, sprite: '...', ...etc... };`
> * Requirements re **OOP**:
>   
>   * [x]  properties common for some classes are generalized into a base class
>     (e.g. there is `Character` base class, which is extended by `Enemy` and `Player` classes)
>   * [x]  class extension is implemented using `Subclass.prototype = Object.create(Superclass.prototype)`,
>     not `Subclass.prototype = new Superclass(params);`;
>     [Useful resource](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Inheritance_and_the_prototype_chain)
>   * [x]  classes do not refer to any global variables, like global variable `player`, which is an instance of `Player` class
>     (referring to global constants and globals provided by the gaming platform like `Resources` is OK);
>     Hint: pass `Player` instance as an argument to every enemy
>   * [x]  Separation of Concerns principle is followed
>     (e.g. `update` method does only rendering and
>     doesn't **contain** any inline code to check e.g. collisions)
>   * [x]  the code is very [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
> * Most common mistakes
>   
>   * [x]  Make sure `target = condition ? valueWhenConditionTrue : valueWhenConditionFalse` is used instead of
>     `condition ? target = valueWhenConditionTrue : target = valueWhenConditionFalse`;
>     [Conditional (ternary) operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator)
> 
> Please, confirm all items above are positive. Please, ask a question if anything above requires clarification.

**Comment by vladk96** [comment]
@OleksiyRudenko I have one question. Please, look at my two functions: `getRandomNumber` and `showMessage`. This functions are in a global scope, but I use them in `Enemy` methods. Is it good or not? Maybe add them to Enemy.prototype but `showMessage` does not related to `Enemy`

**Comment by OleksiyRudenko** [comment]
@vladk96 We can consider `getRandomNumber` a library function, so it's OK to have it in global scope.
`showMessage` would be better a part of game engine (we do not want to intervene into) or `PlayField` object if you had one.

Please, do not fix this.

File: submissions/vladk96/frogger-game/js/app.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    } else {
+        this.x = 0;
+        this.speed = getRandomNumber();
+    }
+
+    this.handlesCollision();
+}
+
+Enemy.prototype.handlesCollision = function() {
+    if (this.player.x < this.x + 60 && this.x < this.player.x + 65 &&
```
Comment (OleksiyRudenko) [inline submissions/vladk96/frogger-game/js/app.js]: Magic  numbers worth being declared as constants.

File: submissions/vladk96/frogger-game/js/app.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+const showMessage = function(boolean) {
+    boolean ? alert('You won!') : alert('You lost!');
+}
+
+const player = new Player(PLAYER_INIT_POS.x, PLAYER_INIT_POS.y);
+const enemy1 = new Enemy(-50, 210, getRandomNumber(), player);
+const enemy2 = new Enemy(-50, 130, getRandomNumber(), player);
+const enemy3 = new Enemy(-50, 60, getRandomNumber(), player);
```
Comment (OleksiyRudenko) [inline submissions/vladk96/frogger-game/js/app.js]: Magic numbers. Worth being declared as constants.

File: submissions/vladk96/frogger-game/js/app.js (hunk 1 above)
Comment (vladk96) [reply submissions/vladk96/frogger-game/js/app.js]: Ok. I will fix it.

**Pull request merged**